src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java

```diff
@@ -239,25 +254,38 @@ public void logSpawn(
       }
       builder.setMnemonic(internalToUnicode(spawn.getMnemonic()));
 
+      boolean warned = false;
       for (ActionInput output : spawn.getOutputFiles()) {
-        Path path = fileSystem.getPath(execRoot.getRelative(output.getExecPath()));
-        if (!output.isDirectory() && !output.isSymlink() && path.isFile()) {
-          builder
-              .addOutputsBuilder()
-              .setOutputId(logFile(output, path, /* inputMetadataProvider= */ null));
-        } else if (output.isDirectory() && path.isDirectory()) {
-          builder
-              .addOutputsBuilder()
-              .setOutputId(logDirectory(output, path, /* inputMetadataProvider= */ null));
-        } else if (output.isSymlink() && path.isSymbolicLink()) {
-          builder
-              .addOutputsBuilder()
-              .setOutputId(logUnresolvedSymlink(output, path, /* inputMetadataProvider= */ null));
-        } else {
-          builder
-              .addOutputsBuilder()
-              .setInvalidOutputPath(internalToUnicode(output.getExecPathString()));
+        var path = fileSystem.getPath(execRoot.getRelative(output.getExecPath()));
+        var outputBuilder = ExecLogEntry.Output.newBuilder();
+        try {
+          if (!output.isDirectory() && !output.isSymlink() && path.isFile()) {
+            outputBuilder.setOutputId(logFile(output, path, /* inputMetadataProvider= */ null));
+          } else if (output.isDirectory() && path.isDirectory()) {
+            outputBuilder.setOutputId(
+                logDirectory(output, path, /* inputMetadataProvider= */ null));
+          } else if (output.isSymlink() && path.isSymbolicLink()) {
+            outputBuilder.setOutputId(
+                logUnresolvedSymlink(output, path, /* inputMetadataProvider= */ null));
+          } else {
+            outputBuilder.setInvalidOutputPath(internalToUnicode(output.getExecPathString()));
+          }
+        } catch (IOException e) {
+          if (!warned) {
+            outputLoggingFailed.set(true);
+            warned = true;
+            logger.at(Level.INFO).log(
+                "Failed to log outputs of %s %s: %s%s",
```

Comment:
Maybe "Failed to log outputs of spawn with mnemonic %s and primary output %s"? Otherwise one might assume that the first output is the one that hit the I/O exception.